File: backend/apps/platform/client_version.py
"""POS client ↔ HQ release alignment.

Tills (Windows, Android, browser terminal) must know when HQ has published a
build that changes operational behaviour. The check is intentionally cheap and
idempotent: clients call it on Sync Centre open and at most once per day, then
surface update_required before stock, cash or clinical work continues.
"""
from __future__ import annotations

from dataclasses import dataclass

from django.utils import timezone

from apps.platform.models import PosRelease


@dataclass(frozen=True)
class ClientVersionStatus:
    platform: str
    client_version: str
    client_build: int
    latest_version: str
    latest_build: int
    update_available: bool
    update_required: bool
    operations_impact: str
    release_notes: str
    checked_at: str
    next_check_after_hours: int = 24

    def as_dict(self) -> dict:
        return {
            "platform": self.platform,
            "client_version": self.client_version,
            "client_build": self.client_build,
            "latest_version": self.latest_version,
            "latest_build": self.latest_build,
            "update_available": self.update_available,
            "update_required": self.update_required,
            "operations_impact": self.operations_impact,
            "release_notes": self.release_notes,
            "checked_at": self.checked_at,
            "next_check_after_hours": self.next_check_after_hours,
        }


def normalize_platform(raw: str) -> str:
    value = (raw or "").strip().upper()
    if value in {PosRelease.Platform.WINDOWS, "WIN", "ELECTRON", "DESKTOP"}:
        return PosRelease.Platform.WINDOWS
    if value in {PosRelease.Platform.ANDROID, "RN", "MOBILE"}:
        return PosRelease.Platform.ANDROID
    if value in {"WEB", "BROWSER", "TERMINAL", "POS_WEB"}:
        # Browser tills track the Windows catalogue until a dedicated web
        # channel exists; operational gates still apply.
        return PosRelease.Platform.WINDOWS
    return value


def evaluate_client_version(
    *,
    platform: str,
    client_version: str = "",
    client_build: int = 0,
) -> ClientVersionStatus:
    resolved = normalize_platform(platform)
    latest = (
        PosRelease.objects.filter(platform=resolved, is_published=True)
        .order_by("-build_number")
        .first()
    )
    now = timezone.now().isoformat()
    if latest is None:
        return ClientVersionStatus(
            platform=resolved or platform,
            client_version=client_version or "",
            client_build=max(0, int(client_build or 0)),
            latest_version="",
            latest_build=0,
            update_available=False,
            update_required=False,
            operations_impact="",
            release_notes="",
            checked_at=now,
        )

    build = max(0, int(client_build or 0))
    update_available = build < latest.build_number
    floor = int(latest.minimum_supported_build or 0)
    update_required = bool(floor and build < floor)
    return ClientVersionStatus(
        platform=resolved,
        client_version=client_version or "",
        client_build=build,
        latest_version=latest.version,
        latest_build=latest.build_number,
        update_available=update_available,
        update_required=update_required,
        operations_impact=latest.operations_impact or "",
        release_notes=latest.release_notes or "",
        checked_at=now,
    )


def client_build_from_request(request) -> tuple[str, str, int]:
    """Read platform / version / build from query params or POS headers."""
    headers = request.headers
    platform = (
        request.query_params.get("platform")
        or headers.get("X-POS-Client-Platform")
        or ""
    )
    version = (
        request.query_params.get("version")
        or headers.get("X-POS-Client-Version")
        or ""
    )
    raw_build = (
        request.query_params.get("build_number")
        or headers.get("X-POS-Client-Build")
        or "0"
    )
    try:
        build = int(raw_build)
    except (TypeError, ValueError):
        build = 0
    return platform, version, build


def reject_if_client_blocked(request) -> dict | None:
    """Return an error payload when a mutating POS call must be refused."""
    platform, version, build = client_build_from_request(request)
    if not platform:
        return None
    status = evaluate_client_version(
        platform=platform, client_version=version, client_build=build
    )
    if not status.update_required:
        return None
    return {
        "detail": (
            f"This till is on build {status.client_build}; HQ requires build "
            f"{status.latest_build} ({status.latest_version}) or newer before "
            "dispensing, cash or clinical operations continue."
        ),
        "code": "POS_CLIENT_UPDATE_REQUIRED",
        "client_version": status.as_dict(),
    }

File: packages/shared/src/operational/clientVersion.ts
/**
 * POS ↔ HQ client version alignment.
 *
 * Tills call this on Sync Centre open and at most once per calendar day so
 * operational upgrades published in HQ (screening, dispensing, cash) surface
 * before the operator keeps working on a stale binary.
 */
import { resolveFetcher } from '../auth/fetcher.js';

export interface PosClientVersionStatus {
  readonly platform: string;
  readonly client_version: string;
  readonly client_build: number;
  readonly latest_version: string;
  readonly latest_build: number;
  readonly update_available: boolean;
  readonly update_required: boolean;
  readonly operations_impact: string;
  readonly release_notes: string;
  readonly checked_at: string;
  readonly next_check_after_hours: number;
}

export type PosClientPlatform = 'WINDOWS' | 'ANDROID' | 'WEB';

export interface CheckPosClientVersionInput {
  readonly platform: PosClientPlatform;
  readonly version: string;
  readonly buildNumber?: number;
  readonly fetcher?: typeof fetch;
  readonly baseUrl?: string;
}

const STORAGE_KEY = 'tibatrace.pos.client-version.v1';
const DAY_MS = 24 * 60 * 60 * 1000;

interface CachedCheck {
  readonly checkedAt: number;
  readonly status: PosClientVersionStatus;
}

function readCache(): CachedCheck | null {
  try {
    if (typeof localStorage === 'undefined') return null;
    const raw = localStorage.getItem(STORAGE_KEY);
    if (!raw) return null;
    return JSON.parse(raw) as CachedCheck;
  } catch {
    return null;
  }
}

function writeCache(status: PosClientVersionStatus): void {
  try {
    if (typeof localStorage === 'undefined') return;
    const payload: CachedCheck = { checkedAt: Date.now(), status };
    localStorage.setItem(STORAGE_KEY, JSON.stringify(payload));
  } catch {
    // Private mode / RN without polyfill — daily check still runs in-memory.
  }
}

/** True when a fresh HQ check is due (or never ran). */
export function isPosClientVersionCheckDue(now = Date.now()): boolean {
  const cached = readCache();
  if (!cached) return true;
  const hours = cached.status.next_check_after_hours || 24;
  return now - cached.checkedAt >= hours * 60 * 60 * 1000 || now - cached.checkedAt >= DAY_MS;
}

export function getCachedPosClientVersion(): PosClientVersionStatus | null {
  return readCache()?.status ?? null;
}

export async function checkPosClientVersion(
  input: CheckPosClientVersionInput,
  options: { readonly force?: boolean } = {},
): Promise<PosClientVersionStatus> {
  if (!options.force && !isPosClientVersionCheckDue()) {
    const cached = getCachedPosClientVersion();
    if (cached) return cached;
  }

  const fetcher = resolveFetcher(input.fetcher);
  const base = (input.baseUrl ?? '').replace(/\/$/, '');
  const params = new URLSearchParams({
    platform: input.platform,
    version: input.version,
    build_number: String(input.buildNumber ?? 0),
  });
  const response = await fetcher(`${base}/api/pos/client-version/?${params.toString()}`, {
    credentials: 'include',
    headers: {
      Accept: 'application/json',
      'X-POS-Client-Platform': input.platform,
      'X-POS-Client-Version': input.version,
      'X-POS-Client-Build': String(input.buildNumber ?? 0),
    },
  });
  if (!response.ok) {
    throw new Error(`Client version check failed (${response.status}).`);
  }
  const status = (await response.json()) as PosClientVersionStatus;
  writeCache(status);
  return status;
}

/** Headers every mutating POS call should carry after a version check. */
export function posClientVersionHeaders(input: {
  readonly platform: PosClientPlatform;
  readonly version: string;
  readonly buildNumber?: number;
}): Record<string, string> {
  return {
    'X-POS-Client-Platform': input.platform,
    'X-POS-Client-Version': input.version,
    'X-POS-Client-Build': String(input.buildNumber ?? 0),
  };
}
